### ISSUE TITLE
`conan install` seemingly hangs, in fact is silently installing with no console output

### ISSUE THREAD

**Issue by agurtovoy** [conan-io/conan#5173]
Conan 1.15.0, Mac OS X 10.14.4 18E226

After terminating a `conan install ..` with Ctrl-C, I ran the command again, and it appeared to hang. Wanting to collect more info, I terminated it again, and then ran `conan install` with CONAN_TRACE_FILE set and discovered that it actually didn't hang, but is *silently* [installing a bunch of dependencies](https://gist.githubusercontent.com/agurtovoy/1341c85af008dc28f694e7874e06ad3f/raw/3f2bccbc7e3b4e5b5c4c9e018472a5b626ef4791/conan.log). The only output I saw while this was happening was this:

```
$ export CONAN_TRACE_FILE=~/conan.log && conan install -p debug ..
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=apple-clang
compiler.cppstd=17
compiler.libcxx=libc++
compiler.version=10.0
os=Macos
os_build=Macos
[options]
[build_requires]
[env]
```

After the install has completed, which took a good 5 minutes, I finally saw the rest of the output:

```
conanfile.py: Installing package
Requirements
    FakeIt/2.0.5@gasuketsu/stable from 'conan-center' - Cache
    boost_algorithm/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_array/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_assert/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_atomic/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_bind/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_chrono/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    boost_concept_check/1.69.0@bincrafters/stable from 'bincrafters' - Cache
    ...
    fmt/5.3.0@bincrafters/stable:2e81740c9856cbfd57822c38999d3b19830708f6 - Cache
    lzma/5.2.4@bincrafters/stable:63363709da4dc3d9b6dcddd47681b79ef69bb2e7 - Missing
    zlib/1.2.11@conan/stable:63363709da4dc3d9b6dcddd47681b79ef69bb2e7 - Missing
    zstd/1.3.5@bincrafters/stable:63363709da4dc3d9b6dcddd47681b79ef69bb2e7 - Missing

FakeIt/2.0.5@gasuketsu/stable: Already installed!
boost_config/1.69.0@bincrafters/stable: Already installed!
boost_predef/1.69.0@bincrafters/stable: Already installed!
boost_preprocessor/1.69.0@bincrafters/stable: Already installed!
bzip2/1.0.6@conan/stable: Already installed!
doctest/2.3.1@bincrafters/stable: Already installed!
fmt/5.3.0@bincrafters/stable: Already installed!
lzma/5.2.4@bincrafters/stable: WARN: Can't find a 'lzma/5.2.4@bincrafters/stable' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, build_type=Debug, compiler=apple-clang, compiler.cppstd=17, compiler.version=10.0, os=Macos
- Options: fPIC=True, shared=False
- Dependencies:
- Package ID: 63363709da4dc3d9b6dcddd47681b79ef69bb2e7

ERROR: Missing prebuilt package for 'lzma/5.2.4@bincrafters/stable'
Try to build it from sources with "--build lzma"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

Not sure what's going on here, but thought I'd report it in case someone else runs into this.

**Comment by lasote** [comment]
Weird, all that should be printed quickly and not buffered. It is the first time I saw. Does it happen always? Can you reproduce it? Is that only happening after a ctrl+c?
Thanks.

**Comment by agurtovoy** [comment]
@lasote It sort of reproduces, at least for this particular project. When I do a clean `install`, it freezes out for about a minute (while the log file shows that it's actually doing stuff), and then the console output catches up with the log file. Sometimes it freezes up in the middle of the install, and then catches up. I thought this was Ctrl-C induced, because that was the first time I saw it, but now I'm not sure.

I nuked my `~/.conan` dir, reinstalled conan and rebooted in a futile attempt to get things back to normal, to no avail. Here's my `conanfile.py`:

```Python
from conans import ConanFile, tools, Meson
import os.path
import os

class Sim( ConanFile ):
    settings = [ "os", "compiler", "build_type", "arch" ]
    generators = "pkg_config"
    requires = \
        "fmt/5.3.0@bincrafters/stable", \
        "FakeIt/2.0.5@gasuketsu/stable", \
        "doctest/2.3.1@bincrafters/stable", \
        "boost_multi_index/1.69.0@bincrafters/stable", \
        "boost_range/1.69.0@bincrafters/stable", \
        "boost_utility/1.69.0@bincrafters/stable"

    def build(self):
        build_dir = "%s_%s" % ( self.settings.build_type, self.settings.arch )
        options = { 'warning_level': 3, 'cpp_std': 'c++17' }        

        meson = Meson(self)
        if self.should_configure:
            meson.configure(source_folder=".", build_folder=build_dir, defs=options)
        else:
            meson.build_folder = build_dir

        if self.should_build:
            meson.build()

        if self.should_test:
            meson.test()
```

Just in case, I'm using iTerm2, Build 3.2.8.

**Comment by lasote** [comment]
Have you tried with the standard Mac terminal? I'm not suggesting you change the iTerm2, just to figure out if it could be related to how the terminal manages the stdout buffers.

**Comment by agurtovoy** [comment]
@lasote I just did, but now it doesn't reproduce in either of them — I tried multiple times :(. I'll try again when/if it comes back. Thanks for the quick responses!

**Comment by agurtovoy** [comment]
@lasote Haha, and 20 minutes later it reproduces again for a release configuration, this time in the standard Mac terminal. Let me know if I can collect any other info that might help you to narrow it down.

**Comment by lasote** [comment]
@jgsogo have you seen this happening on Mac? 
@agurtovoy I have no idea how to narrow it, to be honest :(

**Comment by jgsogo** [comment]
Never for so long (you reported up to  5 minutes), but with huge graphs installing is a very time-consuming operation. After building the graph we check for available binaries given the package_id for each recipe+profile and it involves querying your remotes (if those binaries are not found in the cache). Once they are in the cache, then there is no need to query the remotes and things go really fast (that's the reason why it appeared to you again when you change to the release build).

Using those modular boost time explode, those are ~100 libraries and we are checking for each binary package. According to your trace, each check takes more than 3 seconds (three API calls related to `conaninfo.txt` file) and those sum up your 5 minutes.

The main problem here is **each API call taking 1 second at least**, I don't know if it is always the same or bintray was having a problem at that time. But I've just tried it and I get more or less the same results.

I hope this is not an issue related to Mac or any particular OS/Terminal.

Actions to take:
 * We are working on parallelizing binary downloads, we can extend that effort to these calls inside the graph.
 * Let someone of Bintray know about this, it is a real problem and I agree that one second per petition is too much.
 
OTR.- On my personal projects, I prefer to use the [_monolithic_ boost package](https://bintray.com/conan-community/conan/boost%3Aconan) but of course this is not a solution because the graph can be huge just because it has many libraries.

**Comment by agurtovoy** [comment]
Thanks for the detailed response, @jgsogo! Looking forward to the parallelized downloads. The Bintray API calls are slow, but at least the trace file is updated as they are happening. The console output sometimes is and sometimes not. Are you seeing the same thing on your side?

**Comment by jgsogo** [comment]
The _problem_ with the console is that the first time you run `conan install ...` it shows things like this:

```
boost_unordered/1.69.0@bincrafters/stable: Not found in local cache, looking in remotes...
boost_unordered/1.69.0@bincrafters/stable: Trying with 'conan-center'...
boost_unordered/1.69.0@bincrafters/stable: Trying with 'bincrafters'...
Downloading conanmanifest.txt
[==================================================] 103B/103B        
Downloading conanfile.py
[==================================================] 752B/752B        
Downloading conan_export.tgz
[==================================================] 758B/758B        
Decompressing conan_export.tgz: 100%|██████████| 758/758 [00:00<00:00, 646kB/s]
```

it is looking for the package in the remotes and downloading the recipe, but for the next calls Conan already has the recipe in the cache and knows which remote to use, it is only checking if the binary is available in the remote and during this process, there is no output.

At least this is what happens in my use case, but if this is not yours, please let us know.

Thanks

**Comment by agurtovoy** [comment]
I do experience the pausing you describe, but also from time to time my output is literally just this:

```
$ conan install ..
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=17
compiler.libcxx=libc++
compiler.version=10.0
os=Macos
os_build=Macos
[options]
[build_requires]
[env]
```

... for several minutes (it was five minutes in the extreme case that prompted me to create this issue), while the trace file is showing that dependencies are being downloaded, decompressed, etc.

**Comment by jgsogo** [comment]
Opps! That's a sounds like a different issue. All the download, decompression and so on are printed to the output... It has never happened to me (or I didn't realize about it, most of the time I work just with small projects), and we will need some way to reproduce it and investigate :/

**Comment by agurtovoy** [comment]
@jgsogo Yeah, I never seen it happen until I started pulling in the granular Boost packages, so the depth of the dependency graph might have something to do with this. I don't have the bandwidth to investigate this further at the moment (and it doesn't reproduce consistently), but I'll report back if I run into this when I have a little more time on my hands to do some digging.

**Comment by AntonRechkov** [comment]
I run into similar issue with conan hang in Windows docker container during conan build stage. Downloads and decompress succeed well, but compilation hang for ever. It's sporadic issue

10:14:13  [1] msvc-14.2
10:14:13  [2] msvc-14.2/rls/bst.l-off/bst.l-off/lnk-sttc/pythn-3.6/thrdp-wn32/thrd-mlt/vsblt-hdn
10:14:13  [3] msvc-14.2/rls/bst.l-off/bst.l-off/bld-no/lnk-sttc/pythn-3.6/thrdp-wn32/thrd-mlt/vsblt-hdn
10:14:13  
10:14:13  Component configuration:
10:14:13  
10:14:13      - atomic                   : building
10:14:13      - chrono                   : building
10:14:13      - container                : building
10:14:13      - context                  : building
10:14:13      - contract                 : building
10:14:13      - coroutine                : building
10:14:13      - date_time                : building
10:14:13      - exception                : building
10:14:13      - fiber                    : building
10:14:13      - filesystem               : building
10:14:13      - graph                    : not building
10:14:13      - graph_parallel           : not building
10:14:13      - headers                  : not building
10:14:13      - iostreams                : building
10:14:13      - json                     : building
10:14:13      - locale                   : building
10:14:13      - log                      : building
10:14:13      - math                     : not building
10:14:13      - mpi                      : not building
10:14:13      - nowide                   : building
10:14:13      - program_options          : building
10:14:13      - python                   : not building
10:14:13      - random                   : building
10:14:13      - regex                    : building
10:14:13      - serialization            : building
10:14:13      - stacktrace               : building
10:14:13      - system                   : building
10:14:13      - test                     : not building
10:14:13      - thread                   : building
10:14:13      - timer                    : building
10:14:13      - type_erasure             : not building
10:14:13      - wave                     : not building
10:14:13

**Comment by memsharded** [comment]
@AntonRechkov this could be a different issue, as the above seemed more related to buffering of the output, but it eventually finished and did the output, while your issue is that it really hangs, forever.

This might be a difficult issue to debug, if it cannot be reproduced systematically. Did it happen sometime in a a bare Windows (not docker image)? Is it somewhat possible that it is a resources issue? Sometimes, if the memory is not enough, the behavior can be unexpected, is it a big build? have you tried increasing the memory to the container?

**Comment by AntonRechkov** [comment]
> @AntonRechkov this could be a different issue, as the above seemed more related to buffering of the output, but it eventually finished and did the output, while your issue is that it really hangs, forever.
> 
> This might be a difficult issue to debug, if it cannot be reproduced systematically. Did it happen sometime in a a bare Windows (not docker image)? Is it somewhat possible that it is a resources issue? Sometimes, if the memory is not enough, the behavior can be unexpected, is it a big build? have you tried increasing the memory to the container?

Yes, you right, that's different issue. I increased memory to 16Gb, but it didn't helped. What did help indeed is to decrease number of processors. Issue reproducible with 44 cores, but limiting CPU resources(to 4 cores) makes hang go away.

**Comment by xahon** [comment]
I can't make conan work today at all. It hangs on installation with that in the output window

> inja/3.3.0: Not found in local cache, looking in remotes...
inja/3.3.0: Trying with 'conancenter'...

When i abort the installation and run again it complains about lock files which cannot be eliminated by `conan remove --locks` (#7300)

I tried to reinstall conan with multiple versions ranging from 1.50 to 1.30, with full removal of conan caches and configurations. I also tried to turn off Windows Defender Firewall completely and turn off my VPN. Conan sends a request on https://center.conan.io (Verified via wireshark) but after request it just hangs forever with the output above. Nothing helps

**UPD**
With log level debug i got this

> DEBUG :rest_client_common.py[160]: REST: ping: https://center.conan.io/v1/ping [2022-07-11 01:47:25,616]

https://center.conan.io/v1/ping says 403 Forbidden
